src/lib/tls/asio/asio_async_write_op.h

```diff
@@ -52,24 +54,24 @@ struct AsyncWriteOperation : public AsyncBase<Handler, typename Stream::executor
          {
          m_core.consumeSendBuffer(bytes_transferred);
 
+         if(ec) { m_ec = ec; }
+
          if(!isContinuation)
             {
-            m_ec_store = ec;
             yield m_stream.next_layer().async_write_some(boost::asio::const_buffer(), std::move(*this));
-            ec = m_ec_store;
             }
 
          // the size of the sent TLS record can differ from the size of the payload due to TLS encryption. We need to tell
          // the handler how many bytes of the original data we already processed.
-         this->invoke_now(ec, ec ? 0 : m_plainBytesTransferred);
+         this->invoke_now(m_ec, m_ec ? 0 : m_plainBytesTransferred);
```

Comment:
Why throw out the bytes transferred in case of error?

Reply:
I think we established, that a failed write process should return 0 bytes as consumed. We can't tell the caller that any plaintext bytes were processed, if async_write_some couldn't transfer them.

Reply:
It is entirely possible, depending on the implementation of the `NextLayer` object, for `net_layer().async_write_some()` to return both a non-successful error, and a non-zero bytes transferred.

Reply:
What would we communicate to the handler then? We can't use bytes_transferred, because that's the TLS encrypted payload, which does not need to have any direct correlation with the plaintext size we consumed. And we can't be sure that all bytes were transferred correctly, so `m_plainBytesTransferred` might not be true either.
I had a discussion with Damian last Friday and I think we came to the conclusion that this is how it should be.